cpp/bench/ann/src/common/benchmark.hpp

```diff
@@ -268,24 +267,22 @@ void bench_search(::benchmark::State& state,
         state.SkipWithError(std::string(e.what()));
       }
 
-      auto end = std::chrono::high_resolution_clock::now();
-
-      auto elapsed_seconds = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
       // advance to the next batch
       batch_offset = (batch_offset + n_queries) % query_set_size;
       queries_processed += n_queries;
-      state.SetIterationTime(elapsed_seconds.count());
-      total_time += elapsed_seconds.count();
     }
   }
-  auto end = std::chrono::high_resolution_clock::now();
-  if (state.thread_index() == 0) {
-    auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(end - start).count();
-    state.counters.insert({{"end_to_end", duration}});
-  }
+  cudaDeviceSynchronize();
```

Comment:
Continuing the discussion from here: https://github.com/rapidsai/raft/pull/1920#discussion_r1376622215
Not sure if there is a smaller hammer to guarantee proper `GPU` time.

Reply:
I'm confused by this- why would we synchronize the whole device when the whole goal w/ the multithreading is to allow all the threads to do their work independently?

Reply:
NB: this `cudaDeviceSynchronize();` happens once per invocation of the benchmark case, not each iteration (which is the inner loop on the `state`). This is to ensure the implementation has finished all work in all streams it may have created (e.g. destructors release gpu memory.).

Reply:
Yes, `cudaDeviceSynchronize()` happens outside the benchmark loop. It to ensure that `Latency` is correct even if the benchmarked algorithm does not behave nicely (forget to sync). It does add an overhead, but it is amortized away if we have many benchmark iterations.

Reply:
Many thanks @achirkin for the [detailed explanation](https://github.com/rapidsai/raft/pull/1920#discussion_r1377232547) for the `cuda_timer` and the sync mechanism. It does make sense now.

Reply:
Sorry, I meant to respond to this earlier. Yeah, I was just looking at the diff in my browser and it almost appeared as if the sync was happening on each iteration. Totally fine with doing an initial device sync at the beginning